File: src/main/java/chapter6/Factorial.java
package chapter6;

public class Factorial {

  public static void main(String[] args) {
    System.out.println(factorial(0));
  }

//  private static int factorial(int n) {
//    if (n == 0) {
//      return 1;
//    } else {
//      return n * factorial(n - 1);
//    }
//  }

//  private static int factorial(int n) {
//    if (n <= 1) {
//      return 1;
//    } else {
//      return n * factorial(n - 1);
//    }
//  }

  private static int factorial(int n) {
    int result = 1;

    while (n > 0) {
      result *= n;
      n--;
    }
    return result;
  }

}

File: src/main/kotlin/chapter5/Factorial.kt
package chapter5

fun main() {
  val n = 4
  val factorial = Factorial()
  val result = factorial.calc(n)
  println("Factorial of $n is $result")
}

class Factorial {
  fun calc(n: Int): Int {
    return when (n) {
      1 -> 1
      else -> n * calc(n - 1)
    }
//    if (n == 1) {
//      return 1
//    } else {
//      return n * calc(n - 1)
//    }
  }

}
